src/vision/status_quo/barbara_bridges_sync_and_async.md

```diff
@@ -0,0 +1,245 @@
+
+She decides to try debugging. She fires up a debugger but finds it is isn't really giving her useful information about what is stuck (she has [basically the same problems that Alan has](https://rust-lang.github.io/wg-async-foundations/vision/status_quo/alan_tries_to_debug_a_hang.html)). [She wishes she could get insight into tokio's state.](https://rust-lang.github.io/wg-async-foundations/vision/status_quo/barbara_wants_async_insights.html)
+
+Frustrated, she starts reading the tokio docs more closely and she realizes that `tokio` runtimes offer their own `block_on` method. "What the hey," she thinks, "let's see if that works any better." She changes the `aggregate` function to use tokio's `block_on`:
+
+```rust=
+fn block_on<O>(f: impl Future<Output = O>) -> O {
+    let rt  = tokio::runtime::Runtime::new().unwrap();
+    rt.block_on(f)
+}
+
+fn aggregate(urls: &[Url]) -> Vec<Data> {
+    urls
+        .iter()
+        .map(|url| block_on(do_web_request(url)))
+        .collect()
+}
+```
+
+The good news is that the deadlock is gone. The bad news is that now she is getting a panic:
+
+> thread 'main' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.'
+
+"Well," she thinks, "I could use that `Handle` API to get the current runtime, maybe that will work?"
+
+```rust
+fn aggregate(urls: &[&str]) -> Vec<String> {
+    let handle = Handle::current();
+    urls.iter()
+        .map(|url| handle.block_on(do_web_request(url)))
+        .collect()
+}
+```
+
+But it seems to give her the same panic. 
+
+### Trying out `spawn_blocking`
+
+Reading more into this problem, she realizes she is supposed to be using `spawn_blocking`. She tries replacing `block_on` with `tokio::task::spawn_blocking`:
+
+```rust=
+fn aggregate(urls: &[Url]) {
+    let data: Vec<Data> = 
+        urls
+            .iter()
+            .map(|url| tokio::task::spawn_blocking(move || do_web_request(url)))
+            .collect();
+}
+```
+
+
+but now she gets a type error again:
+
+```
+error[E0277]: a value of type `Vec<Data>` cannot be built from an iterator over elements of type `tokio::task::JoinHandle<impl futures::Future>`
+  --> src/main.rs:22:14
+   |
+22 |             .collect();
+   |              ^^^^^^^ value of type `Vec<Data>` cannot be built from `std::iter::Iterator<Item=tokio::task::JoinHandle<impl futures::Future>>`
+   |
+   = help: the trait `FromIterator<tokio::task::JoinHandle<impl futures::Future>>` is not implemented for `Vec<Data>`
+```
+
+Of course! `spawn_blocking`, like `map`, just takes a regular closure, not an async closure. She's getting a bit frustrated now. "Well," she thinks, "I can use `spawn` to get into an async context!" So she adds a call to `spawn` inside the `spawn_blocking` closure:
```

Comment:
```suggestion
Of course! `spawn_blocking`, like `map`, just takes a regular closure, not an async closure since they're not supported in Rust. She's getting a bit frustrated now. "Well," she thinks, "I can use `spawn` to get into an async context!" So she adds a call to `spawn` inside the `spawn_blocking` closure:
```